### DISCUSSION TITLE
Pass authorizationParams to Google and get them back?

### DISCUSSION THREAD

**Discussion by daniel-meow** [nextauthjs/next-auth#6275] [Help]
Hello my dear developer!

I'm calling for your help. It's my first time working with NextAuth, please be a bit lenient with me. :) 

I’m using 'next-auth' to let users sign up with google. This works. I get the `user` and `account` data back from google inside the `signIn` callback in the `[...nextauth].ts` file. With this information I create a user profile on my backend service.

![image](https://user-images.githubusercontent.com/47478904/210514432-b9a10041-411b-4c70-967d-d4f892dfa69c.png)
`Snippet 1: perfectly working signIn callback with next-auth`

However I have two different types of users: `athletes` and `organisers`. They use separate forms to sign up on the same website. I would like to let both `athletes` and `organisers` sign up with google. But how do I get this information to the signIn callback from "Snipped 1"?

My approach was to pass this information through the `authorizationParams` in the signIn method.

![image](https://user-images.githubusercontent.com/47478904/210514675-1fa5f480-9469-4412-94a8-348979d8ee41.png)
`Snippet 2: Button that passes the userType as authorizationParams`

This now brings me the userType at least to the URL as a parameter in the `oauth consent screen` . But if I select an account now: How I can work with this userType now in my signIn callback? How do I get this parameter back to my application?

![image](https://user-images.githubusercontent.com/47478904/210514810-01b10b51-bbcb-4e5d-b9e9-00edfc538d89.png)
`Snipped 3: Google oauth consent`

Any help would be so much appreciated. I googled the half web for an answer and couldn't find an answer. Maybe I'm searching for the wrong things. I'm lost. Help me. Please :D

**Comment by khuezy** [comment]
@ThangHuuVu Do you have any advice on this? I've read through a few threads around this topic but none of them have a working solution. As Daniel mentioned above, the `custom parameters` are being sent do the `accounts.google.com/signin/oauth/consent?...` url, but are missing from the `/api/auth/callback/google` parameter.

We'd like to be able to access that custom parameter is the `jwt` callback.

**Reply by daniel-meow** [reply]
Hi @khuezy!

I can give you a quick update on how we solved the problem. Disclaimer: It's not the solution we want, but it's the only solution we could find. 

Instead of working with the `custom parameters`, we set a cookie to identify the userType later in NextAuth. 

```
        onClick={() => {
          const expires = dayjs().add(15, 'm').toDate();
          document.cookie = `userType=${userType};path=/;expires=${expires}`;
          signIn(provider.id);
        }}
```

Later in NextAuth, we receive the cookie in the following way.


```
export default function auth(req: NextApiRequest, res: NextApiResponse): ReturnType<typeof NextAuth> {
  return NextAuth(req, res, nextAuthOptions(req, res));
}
```

```
function nextAuthOptions(req: NextApiRequest, res: NextApiResponse): NextAuthOptions {
 ...
 ...
  callbacks: {
    async signIn({ user, account }) {
       ...
       ...
       body: JSON.stringify({ account, user: { ...user, roles: req.cookies.userType } }),
    }
  }
}

```

Hope this might help you until someone finds a better solution!

**Reply by khuezy** [reply]
Thanks for your response and example! Does the cookie automatically delete itself after it expires.

My solution around this was to cache the custom parameter in a local object keyed by the csrf token and then referencing it in the jwt callback.

**Reply by daniel-meow** [reply]
I think the deletion of the (expired) cookies is handled by the browser, so it may differ from browser to browser and I can't give you a 100% answer to your question.

However, as far as I know: Yes, typically browsers automatically delete expired cookies from user's device :)

**Reply by khuezy** [reply]
Turns out my caching solution doesn't work in a production environment (ie lambda). I switched to cookies and it's a great solution imo.
I had to call `toUTCString()` on the `expires`, otherwise it was invalid and the cookie wouldn't get set.
The one caveat is that you have to wrap the authOptions in a closure to access the `req`
